integration-tests/src/tests/client/block_corruption.rs

```diff
@@ -99,3 +104,175 @@ fn change_shard_id_to_invalid() {
         }
     }
 }
+
+/// We ensure there are SOME change between these blocks.
+/// We check that they are NOT in fields that can be mutated and still pass validation by design.
+fn is_breaking_block_change(original: &Block, corrupt: &Block) -> bool {
+    original.header().latest_protocol_version() == corrupt.header().latest_protocol_version()
+        && original.header().block_ordinal() == corrupt.header().block_ordinal()
+        && original.header().timestamp() == corrupt.header().timestamp()
+        && original != corrupt
+}
+
+/// Each block contains one 'send' transaction.
+/// First three blocks are produced without changes.
+/// For the fourth block we are calculating two versions – correct and corrupt.
+/// Corrupt block is produced by
+/// - serializing correct block
+/// - flipping one bit
+/// - deserializing resulting string
+/// - resigning the resulting block
+///  If the corrupt block can be parsed, we are trying to process it and expecting it to fail.
+///  If it successfully fails, we are processing the correct block and expecting it to be ok.
+///
+/// Returns `Ok(processing_error)` if corrupt block was parsed and its processing failed.
+/// Returns `Ok(reason)` if corrupt block wasn't parsed or had changes that are not breaking by design.
+/// Returns `Err(reason)` if corrupt block was processed or correct block wasn't processed afterwards.
+fn check_process_flipped_block_fails_on_bit(
+    corrupted_bit_idx: usize,
+) -> Result<anyhow::Error, anyhow::Error> {
+    let epoch_length = 5000000;
+    let mut genesis = Genesis::test(vec!["test0".parse().unwrap(), "test1".parse().unwrap()], 1);
+    genesis.config.epoch_length = epoch_length;
+    let mut env = TestEnv::builder(ChainGenesis::new(&genesis))
+        .real_epoch_managers(&genesis.config)
+        .nightshade_runtimes(&genesis)
+        .build();
+
+    let mut last_block = env.clients[0].chain.get_block_by_height(0).unwrap();
+
+    let mid_height = 3;
+    for h in 1..=mid_height {
+        let txs = create_tx_load(h, &last_block);
+        for tx in txs {
+            assert_eq!(env.clients[0].process_tx(tx, false, false), ProcessTxResponse::ValidTx);
+        }
+
+        let block = env.clients[0].produce_block(h).unwrap().unwrap();
+        env.process_block(0, block.clone(), Provenance::PRODUCED);
+        last_block = block;
+    }
+
+    let block_len = borsh::to_vec(&last_block).unwrap().len();
+    if corrupted_bit_idx >= block_len * 8 {
+        return Ok(anyhow::anyhow!("End"));
+    }
```

Comment:
nit: Can you perform that check on the last_block, rather than the correct_block. Then you can also skip the modulo business.